rust/private/clippy.bzl

```diff
@@ -0,0 +1,167 @@
+# Copyright 2020 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load(
+    "@io_bazel_rules_rust//rust:private/rustc.bzl",
+    "CrateInfo",
+    "collect_deps",
+    "collect_inputs",
+    "construct_arguments",
+    "construct_compile_command",
+)
+load(
+    "@io_bazel_rules_rust//rust:private/rust.bzl",
+    "crate_root_src",
+    "get_edition"
+)
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+_rust_extensions = [
+    "rs",
+]
+
+def _is_rust_target(srcs):
+    return any([src.extension in _rust_extensions for src in srcs])
+
+def _all_sources(target, rule):
+    srcs = []
+    if "srcs" in dir(rule.attr):
+        srcs += [f for src in rule.attr.srcs for f in src.files.to_list()]
```

Comment:
Extract `src.files.to_list()` to avoid converting twice a hash set

Reply:
So if I expand this out from list comprehension form, we have the following:

```
if "src" in dir(rule.attr):
    for src in rule.attr.srcs:
        for f in src.files.to_list():
            srcs += [f]
```

I can't really pull out `src.files.to_list()`, since this must be accessed on the inner iteration of `rule.attr.srcs`.

That being said, I think it's only being evaluated once? As an example:

```
def foo():
    print("foo")
    return [1, 2, 3]

bar = [f for f in foo()]
```

Only invokes "foo" a single time.

Reply:
Sorry I was reading this incorrectly.